File: routine-api/src/main/java/at/enfilo/def/routine/ClientRoutine.java
package at.enfilo.def.routine;

import at.enfilo.def.communication.dto.ServiceEndpointDTO;
import at.enfilo.def.communication.exception.ClientCommunicationException;
import at.enfilo.def.communication.exception.ClientCreationException;
import at.enfilo.def.execlogic.api.ExecLogicServiceClientFactory;
import at.enfilo.def.execlogic.api.IExecLogicServiceClient;
import at.enfilo.def.parameterserver.api.IParameterServerServiceClient;
import at.enfilo.def.parameterserver.api.client.ParameterServerServiceClientFactory;
import at.enfilo.def.parameterserver.api.protocol.PrimitiveProtocolParser;
import at.enfilo.def.routine.api.Command;
import at.enfilo.def.routine.api.Order;
import at.enfilo.def.routine.api.Result;
import at.enfilo.def.transfer.dto.ParameterProtocol;
import at.enfilo.def.transfer.dto.ProgramDTO;
import at.enfilo.def.transfer.dto.ResourceDTO;
import org.apache.thrift.TException;
import org.apache.thrift.TSerializer;

import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.Map;

public abstract class ClientRoutine extends AbstractRoutine {

    private ProgramDTO program;
    private IParameterServerServiceClient parameterServerServiceClient;
    private Map<String, ResourceDTO> results = new HashMap<>();
    private final TSerializer serializer = new TSerializer();

    @Override
    protected void runRoutine() {
        // fetch pId and client
        try {
            program = getProgram();
            IExecLogicServiceClient client = getServiceClient();
            parameterServerServiceClient = getParameterServerClient();
            results = new HashMap<>();

            routine(program.getId(), client);
            setResult();
        } catch (TException | AccessParameterException e) {
            throw new RoutineException(e);
        }
    }

    /**
     * ClientRoutine
     * Real implementation.
     *
     * @throws RoutineException
     */
    protected abstract void routine(String pId, IExecLogicServiceClient client) throws RoutineException;

    private void setResult() throws TException {
        log(LogLevel.DEBUG, "Storing results.");
        program.setResults(results);
        byte[] data = serializer.serialize(program);
        Result result = new Result(0, "PROGRAM", null, ByteBuffer.wrap(data));
        Order order = new Order(Command.SEND_RESULT, Integer.toString(1));
        ctrl.store(order);
        ctrl.store(result);
        log(LogLevel.INFO, "Client routine results stored.");
    }

    /**
     * Entry point for ClientRoutine
     * @param args
     */
    public static void main(String[] args) {
        if (args.length != 3) {
            String errMsg = "3 Arguments needed: <routineName> <inPipe> <ctrlPipe>";
            throw new RoutineException(errMsg);
        } else {
            try {
                // Instantiate routine and run
                AbstractRoutine routine = createRoutine(args[0], args[1], null, args[2]);
                routine.run();
            } catch (Exception e){
                System.err.println(String.format("%s - Error while running ClientRoutine: %s", LogLevel.ERROR, e.getMessage()));
                throw new RoutineException(e);
            }
        }
    }

    private ProgramDTO getProgram() throws AccessParameterException {
        log(LogLevel.DEBUG, String.format("Requesting program id."));
        Order getPId = new Order(Command.GET_PARAMETER, "program");
        try {
            ctrl.store(getPId);
            ProgramDTO program = in.read(ProgramDTO.class.newInstance());
            log(LogLevel.DEBUG, String.format("Received program id %s.", program.getId()));
            return program;
        } catch (IllegalAccessException | InstantiationException | TException e) {
            throw new AccessParameterException(e);
        }
    }

    private IExecLogicServiceClient getServiceClient() throws AccessParameterException {
        log(LogLevel.DEBUG, String.format("Requesting service client."));
        Order getClient = new Order(Command.GET_PARAMETER, "serviceEndpoint");
        try {
            ctrl.store(getClient);
            ServiceEndpointDTO endpoint = in.read(ServiceEndpointDTO.class.newInstance());
            IExecLogicServiceClient client = new ExecLogicServiceClientFactory().createClient(endpoint);
            return client;
        } catch (IllegalAccessException | InstantiationException | TException | ClientCreationException e) {
            throw new AccessParameterException(e);
        }
    }

    private IParameterServerServiceClient getParameterServerClient() throws AccessParameterException {
        log(LogLevel.DEBUG, "Requesting parameter server client.");
        Order getClient = new Order(Command.GET_PARAMETER, "parameterServerEndpoint");
        try {
            ctrl.store(getClient);
            ServiceEndpointDTO endpoint = in.read(ServiceEndpointDTO.class.newInstance());
            return new ParameterServerServiceClientFactory().createClient(endpoint);
        } catch (IllegalAccessException | InstantiationException | TException | ClientCreationException e) {
            throw new AccessParameterException(e);
        }
    }

    protected void addToResults(String key, ResourceDTO resource) {
        if (key != null) {
            results.put(key, resource);
        }
    }

    public float[] getFloatParameter(String name) throws AccessParameterException {
        try {
            ResourceDTO resourceDTO = parameterServerServiceClient.getParameter(program.getId(), name, ParameterProtocol.PRIMITIVE).get();
            return (float[]) new PrimitiveProtocolParser().decode(resourceDTO);
        } catch (Exception e) {
            throw new AccessParameterException(e);
        }
    }

    public String setParameter(String name, float[] data) throws AccessParameterException {
        try {
            ResourceDTO resourceDTO = new PrimitiveProtocolParser().encode(data, "float");
            return parameterServerServiceClient.setParameter(program.getId(), name, resourceDTO, ParameterProtocol.PRIMITIVE).get();
        } catch (Exception e) {
            throw new AccessParameterException(e);
        }
    }

    public String deleteParameter(String name) throws ClientCommunicationException {
        try {
            return parameterServerServiceClient.deleteParameter(program.getId(), name).get();
        } catch (Exception e) {
            throw new ClientCommunicationException(e);
        }
    }

    public String deleteAllParameters() throws ClientCommunicationException {
        try {
            return parameterServerServiceClient.deleteAllParameters(program.getId()).get();
        } catch (Exception e) {
            throw new ClientCommunicationException(e);
        }
    }

    public String addToParameter(String name, float[] data) throws AccessParameterException {
        try {
            ResourceDTO resourceDTO = new PrimitiveProtocolParser().encode(data, "float");
            return parameterServerServiceClient.addToParameter(program.getId(), name, resourceDTO, ParameterProtocol.PRIMITIVE).get();
        } catch (Exception e) {
            throw new AccessParameterException(e);
        }
    }
}

File: apis/csharp/routine-api/routine/ClientRoutine.cs
using System;
using System.IO;
using System.Collections.Generic;
using common.execlogic;
using common.parameterserver;
using common.thrift;
using routine_api.exception;
using routine_api.util;
using Thrift;

namespace routine_api.routine
{
    public abstract class ClientRoutine : AbstractRoutine
    {
        private ProgramDTO program;
        private ExecLogicClient client;
        private ParameterServerCommunicator parameterServerCommunicator;
        private Dictionary<String, ResourceDTO> results;

        private void SetResult()
        {
            Log(LogLevel.Debug, "Storing results.");
            program.Results = results;
            byte[] data = ThriftSerializer.Serialize(program);
            Result result = new Result
            {
                Seq = 0,
                Key = "PROGRAM",
                Url = null,
                Data = data
            };
            Order order = new Order
            {
                Command = Command.SEND_RESULT,
                Value = "1".ToString()
            };
            CtrlPipe.Store(order);
            CtrlPipe.Store(result);
            Log(LogLevel.Debug, "Client routine results stored.");
        }

        protected override void RunRoutine()
        {
            try
            {
                this.results = new Dictionary<String, ResourceDTO>();
                this.program = GetProgram();
                this.client = GetServiceClient();
                ParameterServerClient parameterServerClient = GetParameterServerClient();
                this.parameterServerCommunicator = new ParameterServerCommunicator(parameterServerClient, this.program.Id);

                Routine(this.program.Id, this.client);
                SetResult();
            }
            catch (Exception e) when (e is TException || e is IOException)
            {
                throw new RoutineException(e);
            }
        }

        protected abstract void Routine(string pId, ExecLogicClient client);

        private ProgramDTO GetProgram()
        {
            Log(LogLevel.Debug, "Requesting program.");
            Order getProgram = new Order
            {
                Command = Command.GET_PARAMETER,
                Value = "program"
            };
            try
            {
                CtrlPipe.Store(getProgram);
                ProgramDTO fetchedProgram = InPipe.Read(Activator.CreateInstance<ProgramDTO>());
                Log(LogLevel.Debug, $"Received: {fetchedProgram}");
                return fetchedProgram;
            }
            catch (Exception e)
            {
                throw new AccessParameterException(e);
            }
        }

        private ExecLogicClient GetServiceClient()
        {
            Log(LogLevel.Debug, "Requesting service client.");
            Order getClient = new Order
            {
                Command = Command.GET_PARAMETER,
                Value = "serviceEndpoint"
            };
            try
            {
                CtrlPipe.Store(getClient);
                ServiceEndpointDTO fetchedEndpoint = InPipe.Read(Activator.CreateInstance<ServiceEndpointDTO>());
                Log(LogLevel.Debug, $"Received: {fetchedEndpoint}");
                ExecLogicClient createdClient = new ExecLogicClient(fetchedEndpoint);
                return createdClient;
            }
            catch (Exception e)
            {
                throw new AccessParameterException(e);
            }
        }

        private ParameterServerClient GetParameterServerClient()
        {
            Log(LogLevel.Debug, "Requesting parameter server client.");
            Order getClient = new Order
            {
                Command = Command.GET_PARAMETER,
                Value = "parameterServerEndpoint"
            };
            try
            {
                CtrlPipe.Store(getClient);
                ServiceEndpointDTO fetchedEndpoint = InPipe.Read(Activator.CreateInstance<ServiceEndpointDTO>());
                Log(LogLevel.Debug, $"Received: {fetchedEndpoint}");
                ParameterServerClient createdClient = new ParameterServerClient(fetchedEndpoint);
                return createdClient;
            }
            catch (Exception e)
            {
                throw new AccessParameterException(e);
            }
        }

        protected void AddToResults(String key, ResourceDTO resource)
        {
            if (key != null)
            {
                results.Add(key, resource);
            }
        }

        /// <inheritdoc />
        /// <summary>
        /// Performs a StoreRoutine specific parsing of parameters and setup.
        /// </summary>
        protected override void Setup(string[] args)
        {
            if (args.Length != 4)
            {
                throw new RoutineException("Wrong Number of Arguments for Client Routine. Usage: <routineName> <inPipe> <ctrlPipe>");
            }

            string inPipeName = args[2];
            string ctrlPipeName = args[3];

            InPipeName = "null".Equals(inPipeName, StringComparison.InvariantCultureIgnoreCase)
                ? null
                : inPipeName;
            OutPipeName = null;
            CtrlPipeName = "null".Equals(ctrlPipeName, StringComparison.InvariantCultureIgnoreCase)
                ? null
                : ctrlPipeName;
        }


        protected String DeleteParameter(String name)
        {
            return this.parameterServerCommunicator.DeleteParameter(name);
        }

        protected String DeleteAllParameters()
        {
            return this.parameterServerCommunicator.DeleteAllParameters();
        }

        // BYTE
        protected byte[] GetByteParameter(String name)
        {
            return this.parameterServerCommunicator.GetByteParameter(name);
        }

        protected String SetParameter(String name, byte[] data)
        {
            return this.parameterServerCommunicator.SetParameter(name, data);
        }

        protected String AddToParameter(String name, byte[] data)
        {
            return this.parameterServerCommunicator.AddToParameter(name, data);
        }

        // SHORT
        protected short[] GetShortParameter(String name)
        {
            return this.parameterServerCommunicator.GetShortParameter(name);
        }

        protected String SetParameter(String name, short[] data)
        {
            return this.parameterServerCommunicator.SetParameter(name, data);
        }

        protected String AddToParameter(String name, short[] data)
        {
            return this.parameterServerCommunicator.AddToParameter(name, data);
        }

        // INT
        protected int[] GetIntParameter(String name)
        {
            return this.parameterServerCommunicator.GetIntParameter(name);
        }

        protected String SetParameter(String name, int[] data)
        {
            return this.parameterServerCommunicator.SetParameter(name, data);
        }

        protected String AddToParameter(String name, int[] data)
        {
            return this.parameterServerCommunicator.AddToParameter(name, data);
        }

        // LONG
        protected long[] GetLongParameter(String name)
        {
            return this.parameterServerCommunicator.GetLongParameter(name);
        }

        protected String SetParameter(String name, long[] data)
        {
            return this.parameterServerCommunicator.SetParameter(name, data);
        }

        protected String AddToParameter(String name, long[] data)
        {
            return this.parameterServerCommunicator.AddToParameter(name, data);
        }

        // FLOAT
        protected float[] GetFloatParameter(String name)
        {
            return this.parameterServerCommunicator.GetFloatParameter(name);
        }

        protected String SetParameter(String name, float[] data)
        {
            return this.parameterServerCommunicator.SetParameter(name, data);
        }

        protected String AddToParameter(String name, float[] data)
        {
            return this.parameterServerCommunicator.AddToParameter(name, data);
        }

        // DOUBLE
        protected double[] GetDoubleParameter(String name)
        {
            return this.parameterServerCommunicator.GetDoubleParameter(name);
        }

        protected String SetParameter(String name, double[] data)
        {
            return this.parameterServerCommunicator.SetParameter(name, data);
        }

        protected String AddToParameter(String name, double[] data)
        {
            return this.parameterServerCommunicator.AddToParameter(name, data);
        }

        // CHAR
        protected char[] GetCharParameter(String name)
        {
            return this.parameterServerCommunicator.GetCharParameter(name);
        }

        protected String SetParameter(String name, char[] data)
        {
            return this.SetParameter(name, data);
        }

        protected String AddToParameter(String name, char[] data)
        {
            return this.AddToParameter(name, data);
        }

        // BOOL
        protected bool[] GetBoolParameter(String name)
        {
            return this.GetBoolParameter(name);
        }

        protected String SetParameter(String name, bool[] data)
        {
            return this.SetParameter(name, data);
        }

        protected String AddToParameter(String name, bool[] data)
        {
            return this.AddToParameter(name, data);
        }
    }
}
